New article: invariant sigils, part 1

cxw42 opened the pull request:
The article referenced in https://github.com/tpf/perldotcom/issues/205#issuecomment-533826834 .

@dnmfarrell This has turned out to be more of a research project than I expected!  I have spent some quality time deep in perly.y and toke.c and have found the experience very enlightening.  However, it does mean that I can't squeeze everything into 1,000 words.  I propose a two-part series, this PR being part 1.

I look forward to your feedback!  Please be as merciless as you see fit :dagger: :D  .

**Edit** I have read the contributing guidelines, style guide, and tutorial.  I have run the text through the MS Word checker and through Hemingway.  However, I do not think this is a final version.  Could you give me some high-level pointers on editing the article to meet perl.com's standards?  Thank you!

dnmfarrell commented:
Hey @cxw42 ,

Thanks for this - I'll review today and provide feedback.

dnmfarrell commented on content/article/my-perl-wishlist-invariant-sigils-part-1.md:
````diff
@@ -0,0 +1,181 @@
+Pop quiz!  Q: What was my mistake in this line?
+
+```perl
+is(%HASH{answer}, 'forty-two', q[%HASH properly filled]);
````
How about simplifying this a bit to:

    is %HASH{answer}, 'forty-two', '%HASH properly filled';

dnmfarrell commented on content/article/my-perl-wishlist-invariant-sigils-part-1.md:
```diff
@@ -0,0 +1,181 @@
+| Sigil | No brackets                                     | `[ ]` (array access)                                                                              | `{ }` (hash access)                                                             |
+|-------|-------------------------------------------------|---------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------|
+| `$`   | `$z`: a scalar, i.e., a single value            | `$z[0]`: the first element of `@z`                                                                | `$z{0}`: the value in `%z` at key `"0"`                                         |
+| `@`   | `@z`: An array, i.e., a list of value(s)        | `@z[0]`: a list of one element from `@z` (an "array slice")                                       | `@z{0}`: the list containing the value `$z{0}` from hash `%z`                   |
```
I like how you've examined the different syntax examples. One thing that could make the slices easier to grok is taking multiple elements in the slice. E.g:

    `@z[0, 1]`

Which makes them more distinct from the single element access syntax.

cxw42 replied:
Please see the new version once I update the PR and let me know if it works better.

dnmfarrell commented on content/article/my-perl-wishlist-invariant-sigils-part-1.md:
```diff
@@ -0,0 +1,181 @@
+
+The adverb (e.g., `:kv`) goes in postfix position, immediately
+after the brackets or braces.  Following this model,
+slices would look like `@z[1,2,3]:l` or `%hash{a}:kv`.  (I propose `:l`
```
It *might* help to reword this slightly to be:

> I propose `:l` as in list instead of `:v` which can return a scalar or a list

cxw42 replied:
Revised

dnmfarrell commented on content/article/my-perl-wishlist-invariant-sigils-part-1.md:
```diff
@@ -0,0 +1,181 @@
+| List of values          | `@z`: An array, i.e., a list of value(s)        | `@z@[0]` / `@z[0]:l`: a list including the first element of `@z`             | `%z@{0}` / `%z{0}:l`: A list including the `"0"` element of `%z`                        |
+| List of key/value pairs | `%z`: A hash, i.e., a list of key/value pair(s) | `@z%[0]` / `@z[0]:kv`: a list of the index `0` and the first element of `@z` | `%z%{0}` / `%z{0}:kv`: a list of the key `"0"` and the value of element `"0"` of `%z` |
+
+You can't always get what you want
```
I love how you dug into the parser and what's possible here.

cxw42 replied:
Just wait until part 2 :D

dnmfarrell commented on content/article/my-perl-wishlist-invariant-sigils-part-1.md:
In my browser this text shows up. If you change it to an html comment, it will be hidden.

cxw42 replied:
Removed

cxw42 commented:
@dnmfarrell Thank you for your comments!  I have added the second draft.  I added it as a separate commit so that you could see the edits more easily.  I am happy to force-push back to a single commit if you prefer.  Alternatively, if you would be willing to squash when the time comes, I would appreciate it!

I look forward to any further comments or questions you have on Part 1.  Much appreciated!

dnmfarrell commented:
Thanks @cxw42. Merged into master, this will go live shortly

cxw42 commented:
@dnmfarrell Thank you very much indeed!  I will open a PR for Part 2 when it is ready.  I need to do a bit more experimenting, so it probably won't be tomorrow ;) .
